src/translate.rs

```diff
@@ -391,32 +444,58 @@ fn compile_expr(ctxt: &rustast::ExtCtxt, e: &Expr, result_used: bool) -> rustast
 
 		RuleExpr(ref rule_name) => {
 			let func = rustast::str_to_ident(&format!("parse_{}", *rule_name));
-			quote_expr!(ctxt, $func(input, state, pos))
+			let rule = grammar.find_rule(rule_name);
+			match rule {
+				Some(rule) if rule.cached => {
+					let cache_field = rustast::str_to_ident(&format!("{}_cache", *rule_name));
+
+					if cfg!(feature = "trace") {
+						quote_expr!(ctxt, {
+							state.$cache_field.get(&pos).map(|entry| {
+								let (line, col) = pos_to_line(input, pos);
+								match entry {
+									&Matched(..) => println!("[PEG_TRACE] Cached match of rule {} at {}:{} (pos {})", $rule_name, line, col, pos),
+									&Failed => println!("[PEG_TRACE] Cached fail of rule {} at {}:{} (pos {})", $rule_name, line, col, pos),
+								};
+
+								entry.clone()
+							}).unwrap_or_else(|| $func(input, state, pos))
+						})
+					} else {
+						quote_expr!(ctxt, {
+							state.$cache_field.get(&pos).map(|entry| entry.clone()).unwrap_or_else(|| $func(input, state, pos))
```

Comment:
It would be interesting to benchmark performance and code size with the cache lookup in the caller vs at the beginning of the rule function. This can happen later, though.